libethash-cuda/CUDAMiner.cpp

```diff
@@ -457,96 +452,51 @@ void CUDAMiner::search(
             // Wait for the stream complete
             CUDA_SAFE_CALL(cudaStreamSynchronize(stream));
 
-            // refer to the current stream's solution buffer
-            volatile Search_results& buffer(*m_search_buf[current_index]);
-            // See if the stream found any solutions
-            uint32_t found_count = buffer.count;
-            if (found_count)
+            if (shouldStop())
             {
-#ifdef DEV_BUILD
-                submitStart = std::chrono::steady_clock::now();
-#endif
-                // Found one or more solutions, save them for later.
-                // We are in a hurry to restart the stream, so defer
-                // non-critical activities, such as hash rate calculations
-                // and solution submission.
+                m_new_work.store(false, std::memory_order_relaxed);
+                done = true;
+            }
 
-                if (found_count > MAX_SEARCH_RESULTS)
-                    found_count = MAX_SEARCH_RESULTS;
+            // Detect solutions in current stream's solution buffer
+            volatile Search_results& buffer(*m_search_buf[current_index]);
+            uint32_t found_count = std::min((unsigned)buffer.count, MAX_SEARCH_RESULTS);
 
-                // refer to the current stream's save buffer
+            if (found_count)
+            {
+                buffer.count = 0;
+                uint64_t nonce_base = start_nonce - streams_batch_size;
+                
+                // Extract solution and pass to higer level 
+                // using io_service as dispatcher
 
-                // copy the solution to the save buffer for this stream
-                save_buf.count = found_count;
                 for (uint32_t i = 0; i < found_count; i++)
                 {
-                    save_buf.result[i].gid = buffer.result[i].gid;
-                    if (s_noeval)
-                        for (uint32_t j = 0; j < 8; j++)
-                            save_buf.result[i].mix[j] = buffer.result[i].mix[j];
-                }
+                    h256 mix;
+                    uint64_t nonce = nonce_base + buffer.result[i].gid;
+                    memcpy(mix.data(), (void*)&buffer.result[i].mix, sizeof(buffer.result[i].mix));
+                    auto sol =
+                        Solution{nonce, mix, w, done, std::chrono::steady_clock::now(), m_index};
+
+                    if (sol.stale)
+                        cwarn << "Stale solution: " << EthWhite "0x" << toHex(sol.nonce)
+                              << EthReset;
+                    else
+                        cnote << "Solution: " << EthWhite "0x" << toHex(sol.nonce) << EthReset;
 
-                // Ok, solutions are saved. We can now reset the stream's
-                // solution buffer, releasing it for reuse.
-                buffer.count = 0;
+                    g_io_service.post(
+                        m_io_strand.wrap(boost::bind(&Farm::submitProof, &Farm::f(), sol)));
```

Comment:
Bad idea... You're forcing a thread switch to submit.

Reply:
It shouldn't be so.

>This function is used to ask the io_service to execute the given handler, but without allowing the io_service to call the handler from inside this function.
The io_service guarantees that the handler will only be called in a thread in which the run(), run_one(), poll() or poll_one() member functions is currently being invoked.

Actually I'm posting a shared handler and a parameter to a separate thread.

Reply:
I would be reassuring to provide some actual delay measurements, between the time the host detects a solution and the time it gets to the ASIO send.